packages/cf.js/src/provider.ts

```diff
@@ -40,43 +47,150 @@ export class Provider {
-          reject(new Error(`Request timed out: ${requestId}`));
+          reject({
+            type: EventType.ERROR,
+            data: {
+              errorName: "request_timeout",
+              message: `Request timed out: ${JSON.stringify(request)}`
+            }
+          });
           delete this.requestListeners[requestId];
         }
       }, NODE_REQUEST_TIMEOUT);
-      this.nodeProvider.sendMessage({
-        requestId,
-        params,
-        type: methodName
-      });
+      this.nodeProvider.sendMessage(request);
     });
   }
 
   private onNodeMessage(message: Node.Message) {
-    const requestId = (message as Node.MethodResponse).requestId;
-    if (requestId) {
-      if (this.requestListeners[requestId]) {
-        this.requestListeners[requestId](message);
-        delete this.requestListeners[requestId];
-      }
+    const type = message.type;
+    if (Object.values(Node.ErrorType).indexOf(type) !== -1) {
+      this.handleNodeError(message as Node.Error);
+    } else if ((message as Node.MethodResponse).requestId) {
+      this.handleNodeMethodResponse(message as Node.MethodResponse);
     } else {
-      // TODO: notify observers
+      this.handleNodeEvent(message as Node.Event);
+    }
+  }
+
+  private handleNodeError(error: Node.Error) {
+    const requestId = error.requestId;
+    if (requestId && this.requestListeners[requestId]) {
+      this.requestListeners[requestId](error);
+      delete this.requestListeners[requestId];
+    }
+    this.eventEmitter.emit(error.type, error);
+  }
+
+  private handleNodeMethodResponse(response: Node.MethodResponse) {
+    const { requestId } = response;
+    if (this.requestListeners[requestId]) {
+      this.requestListeners[requestId](response);
+      delete this.requestListeners[requestId];
+    } else {
+      const error = {
+        type: EventType.ERROR,
+        data: {
+          errorName: "orphaned_response",
+          message: `Response has no corresponding inflight request: ${JSON.stringify(
+            response
+          )}`
+        }
+      };
+      this.eventEmitter.emit(error.type, error);
+    }
+  }
+
+  private async getOrCreateAppInstance(
+    id: AppInstanceID,
+    info?: AppInstanceInfo
+  ): Promise<AppInstance> {
+    if (!this.appInstances[id]) {
+      let newInfo;
+      if (info) {
+        newInfo = info;
+      } else {
+        const { result } = await this.callNodeMethod(
+          Node.MethodName.GET_APP_INSTANCE_DETAILS,
+          { appInstanceId: id }
+        );
+        newInfo = (result as Node.GetAppInstanceDetailsResult).appInstance;
+      }
+      this.appInstances[id] = new AppInstance(newInfo);
+    }
+    return this.appInstances[id];
+  }
+
+  private async handleNodeEvent(nodeEvent: Node.Event) {
+    let event: CounterfactualEvent;
+    switch (nodeEvent.type) {
+      case Node.EventName.REJECT_INSTALL: {
+        const data = nodeEvent.data as Node.RejectInstallEventData;
```

Comment:
Is type casting here equivalent to
```
const data: Node.RejectInstallEventData = nodeEvent.data
```
?

Reply:
No actually, I tried that but Typescript complained.